streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -0,0 +1,428 @@
+        }
+
+        @Override
+        public void process(final K key, final V value) {
+            if (reverseIteratorImplemented) {
+                processReverse(key, value);
+            } else {
+                processInOrder(key, value);
+            }
+        }
+
+        public void processReverse(final K key, final V value) {
+            if (key == null) {
+                log.warn(
+                        "Skipping record due to null key. value=[{}] topic=[{}] partition=[{}] offset=[{}]",
+                        value, context().topic(), context().partition(), context().offset()
+                );
+                droppedRecordsSensor.record();
+                return;
+            }
+
+            final long timestamp = context().timestamp();
+            observedStreamTime = Math.max(observedStreamTime, timestamp);
+            final long closeTime = observedStreamTime - windows.gracePeriodMs();
+            //store start times of windows we find
+            final HashSet<Long> windowStartTimes = new HashSet<Long>();
+            // aggregate that will go in the current record’s left/right window (if needed)
+            ValueAndTimestamp<Agg> leftWinAgg = null;
+            ValueAndTimestamp<Agg> rightWinAgg = null;
+
+            // flag to help determine if out-of-order record’s right window is non-empty
+            boolean foundLeftFirst = false;
+            //if current record's left/right windows already exist
+            boolean leftWinExists = false;
+            boolean rightWindowExists = false;
+            //to determine if we're creating the previous record's right window, helps with determining empty windows
+            boolean prevRightWinExists = false;
+            boolean prevRightWinAlreadyCreated = false;
+
+            try (
+                    //Fetch all the windows that have a start time between timestamp and timestamp+timeDifference
+                    final KeyValueIterator<Windowed<K>, ValueAndTimestamp<Agg>> iterator = windowStore.fetch(key,
+                            key,
+                            Instant.ofEpochMilli(timestamp - 2 * windows.timeDifference()),
+                            Instant.ofEpochMilli(timestamp + 1))
+            ) {
+                KeyValue<Windowed<K>, ValueAndTimestamp<Agg>> next;
+                //if we've already seen the window with the closest start time to the record
+                boolean foundFirst = false;
+                //if we've already seen the window with the closest end time to the record
+                boolean foundFirstEndTime = false;
+                while (iterator.hasNext()) {
+                    next = iterator.next();
+                    windowStartTimes.add(next.key.window().start());
+
+                    //determine if current record's right window exists
+                    if (next.key.window().start() == timestamp + 1) {
+                        rightWindowExists = true;
+                        continue;
```

Comment:
I'm not sure if it can be moved out of the loop unless we also move a check for if the first is a window whose aggregate we need to update, which is easy to do but gets back to the redundant code that the algorithm had before having one while loop. I did update the comment though

Reply:
>  unless we also move a check for if the first is a window whose aggregate we need to update

Which check? I was just thinking that, since the window starting at record.timestamp + 1 is basically a special case, we can just pull it out of the loop completely. We don't have to update anything since the record doesn't fall into this window, right? Basically just before entering the loop we check `if next.key.window().start() == timestamp + 1` and if so set `rightWinAlreadyCreated` and then skip to the next record

Reply:
Yeah that works unless the first window isn't the right window, in which case we would need to process it (save as the rightWinAgg, update its aggregate if the current record falls into it, etc) before going to the next record at the top of the while loop. It definitely works and is what we had before, it just makes the code a little less clean.

Reply:
Well, if the first window isn't the right window then we just wouldn't call `iterator.next` again, right? So we wouldn't have to do anything at all